.github/workflows/ci-typescript.yml

```
@@ -14,12 +20,231 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
+  detect-changes:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    outputs:
+      global-changes: ${{ steps.filter.outputs.global-changes }}
+      individual-lambdas-changed: ${{ steps.filter.outputs.changes }}
+      any-lambda-changed: ${{ steps.check-lambdas.outputs.any-changed }}
+      lambdas-to-build: ${{ steps.check-lambdas.outputs.lambdas-to-build }}
+      all-typescript-lambdas: ${{ steps.discover-lambdas.outputs.lambdas }}
+    steps:
+      - uses: actions/checkout@v5
+      - name: Discover TypeScript lambdas
+        id: discover-lambdas
+        run: |
+          # Find all directories in handlers/ that contain package.json (TypeScript lambdas)
+          typescript_lambdas=()
+          for dir in handlers/*/; do
+            if [[ -f "$dir/package.json" ]]; then
+              # Extract directory name without handlers/ prefix and trailing /
+              lambda_name=$(basename "$dir")
+              typescript_lambdas+=("$lambda_name")
+            fi
+          done
+          
+          echo "DEBUG: Found ${#typescript_lambdas[@]} TypeScript lambdas: ${typescript_lambdas[*]}"
+          
+          # Convert to compact JSON array using jq - with multiple fallbacks
+          if [ ${#typescript_lambdas[@]} -eq 0 ]; then
+            lambdas_json="[]"
+            echo "INFO: No TypeScript lambdas found"
+          else
+            # Primary method: use jq
+            if command -v jq >/dev/null 2>&1; then
+              lambdas_json=$(printf '%s\n' "${typescript_lambdas[@]}" | jq -R . | jq -s -c .)
+              echo "DEBUG: jq generated: $lambdas_json"
+              
+              # Validate the generated JSON
+              if echo "$lambdas_json" | jq . >/dev/null 2>&1; then
+                echo "SUCCESS: Valid JSON generated with jq"
+              else
+                echo "ERROR: jq generated invalid JSON, using fallback"
+                lambdas_json="[]"
+              fi
+            else
+              echo "ERROR: jq not available, using fallback"
+              lambdas_json="[]"
+            fi
+          fi
+          
+          echo "lambdas=$lambdas_json" >> $GITHUB_OUTPUT
+          echo "INFO: Final output: $lambdas_json"
+      - name: Generate dynamic paths filter
+        id: generate-filter
+        run: |
+          # Get discovered lambdas from the previous step
+          all_typescript_lambdas='${{ steps.discover-lambdas.outputs.lambdas }}'
+          
+          echo "DEBUG: Raw lambdas JSON: $all_typescript_lambdas"
+          
+          # Write base filter config
+          cat > /tmp/filters.yml << 'EOF'
```

might be a bit clearer?
```suggestion
          cat > /tmp/build-trigger-groups.yml << 'EOF'
```